### PR TITLE
add 7TV Emotes (https://7tv.app/)

### PR DESCRIPTION
**Pull request by Nickwasused**
7TV is an alternative to BTTV. 
Their Code is Open Source and can be found here: https://github.com/SevenTV
Their Website can be found here: https://7tv.app/
Privacy Policy of 7TV: https://7tv.app/legal/privacy

### Preview
[](https://www.youtube.com/watch?v=CvjD5yX0eBU)

# To-do

- [x] Fix broken emotes (While testing I noticed that only about 1 out of 500 emotes is broken.)
- [x] Opt-In (Settings)
- [x] Implement Caching (Caching is going to be implemented another time.)

### CODE REVIEW

**Comment by Nickwasused** [comment]
I have added Settings for all three Emote Services (BetterTTV, FFZ and 7TV). BTTV and FFZ are enabled by default like before and 7TV is opt-in.

File: app/src/main/java/com/perflyst/twire/chat/ChatEmoteManager.java [hunk 1]
```diff
@@ -135,6 +161,45 @@ void loadCustomEmotes(EmoteFetchCallback callback) {
...
+                seventvResponsechannel = Service.urlToJSONString(SEVENTV_CHANNEL_URL);
+            }
+
+            // get global emotes
+            JSONArray seventvemotesglobal = new JSONArray(seventvResponseglobal);
+
+            // Read all the emotes
+            for (int i = 0; i < seventvemotesglobal.length(); i++) {
+                Emote emote = To7TV(seventvemotesglobal.getJSONObject(i));
+                emote.setCustomChannelEmote(true);
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/chat/ChatEmoteManager.java:184]: This shouldn't be set to true, it's a global emote.
```suggestion
```

File: app/src/main/java/com/perflyst/twire/chat/ChatManager.java [hunk 2]
```diff
@@ -77,8 +77,17 @@
...
-        mEmoteManager = new ChatEmoteManager(aChannel, aChannelUserId);
         Settings appSettings = new Settings(aContext);
+        JSONObject settings_object = new JSONObject();
+        try {
+            settings_object.put("enable_bttv", appSettings.getChatEmoteBTTV());
+            settings_object.put("enable_ffz", appSettings.getChatEmoteFFZ());
+            settings_object.put("enable_seventv", appSettings.getChatEmoteSEVENTV());
+        } catch (JSONException e) {
+            e.printStackTrace();
+        }
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/chat/ChatManager.java:88]: I feel like using a JSONObject for this causes more hassle than just creating a quick class to store these three variables.

File: app/src/main/java/com/perflyst/twire/chat/ChatManager.java [hunk 3]
```diff
@@ -77,8 +77,17 @@
...
+        Log.d(LOG_TAG, "Chat Emote Settings: " + settings_object.toString());
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/chat/ChatManager.java:89]: If this debug statement is not useful to keep around, be sure to remove it.

**Review by samfundev — CHANGES_REQUESTED** [review]
Noticed a few things that could be improved.

File: app/src/main/java/com/perflyst/twire/chat/ChatManager.java (hunk 3 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/chat/ChatManager.java]: removed in b2b8021e45554adfca7278d49cef05ababaa026d

File: app/src/main/java/com/perflyst/twire/chat/ChatEmoteManager.java (hunk 1 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/chat/ChatEmoteManager.java]: set to false in b2b8021e45554adfca7278d49cef05ababaa026d

File: app/src/main/java/com/perflyst/twire/chat/ChatManager.java (hunk 2 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/chat/ChatManager.java]: I have added the Class "ChatEmoteSettings" in b2b8021e45554adfca7278d49cef05ababaa026d.

File: app/src/main/java/com/perflyst/twire/chat/ChatEmoteManager.java [hunk 4]
```diff
@@ -33,10 +34,12 @@
 
     private final String channelName;
     private final int channelId;
+    private final ChatEmoteSettings emote_settings;
 
-    ChatEmoteManager(String channelName, int channelId) {
+    ChatEmoteManager(String channelName, int channelId, ChatEmoteSettings emotesettings) {
```
Comment (samfundev) [inline app/src/main/java/com/perflyst/twire/chat/ChatEmoteManager.java:39]: Sorry, after reviewing this again I realized that there is an even easier solution. Just pass the settings along from the ChatManager.
```suggestion
    ChatEmoteManager(String channelName, int channelId, Settings settings) {
```

**Review by samfundev — CHANGES_REQUESTED** [review]
Apologizes, but I realized there is an even easier solution that would simplify the code.

File: app/src/main/java/com/perflyst/twire/chat/ChatEmoteManager.java (hunk 4 above)
Comment (Nickwasused) [reply app/src/main/java/com/perflyst/twire/chat/ChatEmoteManager.java]: Done in 9ac85cef6e545d9bc5ea01b033c01d6ca5fa9375

**Pull request merged**